packages/@aws-cdk/aws-sagemaker/lib/endpoint-config.ts

```diff
@@ -0,0 +1,319 @@
+   */
+  public static fromEndpointConfigName(scope: Construct, id: string, endpointConfigName: string): IEndpointConfig {
+    const endpointConfigArn = cdk.Stack.of(scope).formatArn({
+      service: 'sagemaker',
+      resource: 'endpoint-config',
+      resourceName: endpointConfigName,
+    });
+
+    class Import extends cdk.Resource implements IEndpointConfig {
+      public endpointConfigArn = endpointConfigArn;
+      public endpointConfigName = endpointConfigName;
+    }
+
+    return new Import(scope, id, {
+      environmentFromArn: endpointConfigArn,
+    });
+  }
+
+  /**
+   * The ARN of the endpoint configuration.
+   */
+  public readonly endpointConfigArn: string;
+  /**
+   * The name of the endpoint configuration.
+   */
+  public readonly endpointConfigName: string;
+
+  private readonly instanceProductionVariantsByName: { [key: string]: InstanceProductionVariant; } = {};
+
+  constructor(scope: Construct, id: string, props: EndpointConfigProps = {}) {
+    super(scope, id, {
+      physicalName: props.endpointConfigName,
+    });
+
+    (props.instanceProductionVariants || []).map(p => this.addInstanceProductionVariant(p));
+
+    // create the endpoint configuration resource
+    const endpointConfig = new CfnEndpointConfig(this, 'EndpointConfig', {
+      kmsKeyId: (props.encryptionKey) ? props.encryptionKey.keyArn : undefined,
+      endpointConfigName: this.physicalName,
+      productionVariants: cdk.Lazy.any({ produce: () => this.renderInstanceProductionVariants() }),
+    });
+    this.endpointConfigName = this.getResourceNameAttribute(endpointConfig.attrEndpointConfigName);
+    this.endpointConfigArn = this.getResourceArnAttribute(endpointConfig.ref, {
+      service: 'sagemaker',
+      resource: 'endpoint-config',
+      resourceName: this.physicalName,
+    });
+  }
+
+  /**
+   * Add production variant to the endpoint configuration.
+   *
+   * @param props The properties of a production variant to add.
+   */
+  public addInstanceProductionVariant(props: InstanceProductionVariantProps): void {
+    if (props.variantName in this.instanceProductionVariantsByName) {
+      throw new Error(`There is already a Production Variant with name '${props.variantName}'`);
+    }
+    this.validateInstanceProductionVariantProps(props);
+    this.instanceProductionVariantsByName[props.variantName] = {
+      acceleratorType: props.acceleratorType,
+      initialInstanceCount: props.initialInstanceCount || 1,
+      initialVariantWeight: props.initialVariantWeight || 1.0,
+      instanceType: props.instanceType || InstanceType.T2_MEDIUM,
+      modelName: props.model.modelName,
+      variantName: props.variantName,
+    };
+  }
+
+  /**
+   * Get instance production variants associated with endpoint configuration.
+   *
+   * @internal
+   */
+  public get _instanceProductionVariants(): InstanceProductionVariant[] {
+    return Object.values(this.instanceProductionVariantsByName);
+  }
```

Comment:
lets have a conversation about `@internal` public APIs here, and have it apply to the rest of the PR. The two that i'm noticing, although there may be more, is `_instanceProdVariants()` and `_findInstanceProdVariants()`

`_instanceProdVariants()` doesn't seem to be used outside of the construct, so why is it public in the first place?
`_findInstanceProdVariants()` is being used outside the construct in testing only.

I don't really like using this pattern for the express purpose of utilizing the API in testing, as it seems to be the case here. Generally, I'm okay with `@internal` only if it is necessary for a different construct to consume the method. From what I can see, the tests may have to be clunkier, but we can assert on the resulting template itself rather than using this getter and avoid synthing. I'm likely missing something in your reasoning for using `@internal`, so let me know your thought process and see if we can come to a reasonable solution.

Reply:
Pre-empting your (possible) response: is this being used in `Endpoint`? If so, can you explain a bit about how/why?

Reply:
Yes, these are being used by `Endpoint`:
- These `@internal` methods are used by similarly named functions on `Endpoint` named `get instanceProductionVariants ` and `findInstanceProductionVariant` (see [this section of the RFC](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0431-sagemaker-l2-endpoint.md#endpoint-1)).
- This [other section of the RFC](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0431-sagemaker-l2-endpoint.md#what-are-the-drawbacks-of-this-solution) describes the purpose behind these methods.
- The [AutoScaling](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0431-sagemaker-l2-endpoint.md#autoscaling) and [Metrics](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0431-sagemaker-l2-endpoint.md#metrics) sections of the RFC's README includes examples using this feature.

In short, as variants deployed to an endpoint represent an auto-scalable monitor able entity, `Endpoint` inspects the supplied `EndpointConfig`'s variants to offer `metric*` and `autoScale*` APIs with synthesis-time safeguards to ensure that customers are in fact monitoring existing variants.

Reply:
If you'd like, I can rip these `@internal` methods out of this PR and include them in the `Endpoint` PR where their purpose is clearer (as of now, in this PR, they are simply being tested and serve no other purpose).

Reply:
Nope. This is great, and sorry I know I could have found these resources by myself, but I was in a rush.

`@internal` is meant for when we need to reference a construct's methods in other constructs, and we do not wish to expose the API to the consumer. This is that case, so we are good with this method.